<a href="https://colab.research.google.com/github/manigs2007/Credit_Card_Fraud_Detection/blob/main/Credit_card_fraud.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

## Importing the Dependencies

In [1]:
import numpy as np
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score

In [2]:
## Loading the dataset

credit_card_data = pd.read_csv("/content/creditcard.csv")

In [3]:
credit_card_data.head()  ## first 5 rows

Time        V1        V2        V3        V4        V5        V6        V7  \
0   0.0 -1.359807 -0.072781  2.536347  1.378155 -0.338321  0.462388  0.239599   
1   0.0  1.191857  0.266151  0.166480  0.448154  0.060018 -0.082361 -0.078803   
2   1.0 -1.358354 -1.340163  1.773209  0.379780 -0.503198  1.800499  0.791461   
3   1.0 -0.966272 -0.185226  1.792993 -0.863291 -0.010309  1.247203  0.237609   
4   2.0 -1.158233  0.877737  1.548718  0.403034 -0.407193  0.095921  0.592941   

         V8        V9  ...       V21       V22       V23       V24       V25  \
0  0.098698  0.363787  ... -0.018307  0.277838 -0.110474  0.066928  0.128539   
1  0.085102 -0.255425  ... -0.225775 -0.638672  0.101288 -0.339846  0.167170   
2  0.247676 -1.514654  ...  0.247998  0.771679  0.909412 -0.689281 -0.327642   
3  0.377436 -1.387024  ... -0.108300  0.005274 -0.190321 -1.175575  0.647376   
4 -0.270533  0.817739  ... -0.009431  0.798278 -0.137458  0.141267 -0.206010   

        V26       V27       V28  Amount  Class  
0 -0.189115  0.133558 -0.021053  149.62      0  
1  0.125895 -0.008983  0.014724    2.69      0  
2 -0.139097 -0.055353 -0.059752  378.66      0  
3 -0.221929  0.062723  0.061458  123.50      0  
4  0.502292  0.219422  0.215153   69.99      0  

[5 rows x 31 columns]

In [4]:
credit_card_data.shape  ## dataset shape

(284807, 31)

We observe from the dataset that time is present in seconds. The subsequent transactions are that many seconds past the 1st transaction. eg 5th transaction happened 2 seconds past 1st transaction

Since, the data has to be classified because it conatins personal details it is given as V1, V2.. etc. 
Amount is in US Dollars. 
Class is binary coded. Therefore Class = 0 means legit transaction while Class = 1 means fraudulent transaction.

In [5]:
credit_card_data.tail()

Time         V1         V2        V3        V4        V5  \
284802  172786.0 -11.881118  10.071785 -9.834783 -2.066656 -5.364473   
284803  172787.0  -0.732789  -0.055080  2.035030 -0.738589  0.868229   
284804  172788.0   1.919565  -0.301254 -3.249640 -0.557828  2.630515   
284805  172788.0  -0.240440   0.530483  0.702510  0.689799 -0.377961   
284806  172792.0  -0.533413  -0.189733  0.703337 -0.506271 -0.012546   

              V6        V7        V8        V9  ...       V21       V22  \
284802 -2.606837 -4.918215  7.305334  1.914428  ...  0.213454  0.111864   
284803  1.058415  0.024330  0.294869  0.584800  ...  0.214205  0.924384   
284804  3.031260 -0.296827  0.708417  0.432454  ...  0.232045  0.578229   
284805  0.623708 -0.686180  0.679145  0.392087  ...  0.265245  0.800049   
284806 -0.649617  1.577006 -0.414650  0.486180  ...  0.261057  0.643078   

             V23       V24       V25       V26       V27       V28  Amount  \
284802  1.014480 -0.509348  1.436807  0.250034  0.943651  0.823731    0.77   
284803  0.012463 -1.016226 -0.606624 -0.395255  0.068472 -0.053527   24.79   
284804 -0.037501  0.640134  0.265745 -0.087371  0.004455 -0.026561   67.88   
284805 -0.163298  0.123205 -0.569159  0.546668  0.108821  0.104533   10.00   
284806  0.376777  0.008797 -0.473649 -0.818267 -0.002415  0.013649  217.00   

        Class  
284802      0  
284803      0  
284804      0  
284805      0  
284806      0  

[5 rows x 31 columns]

Time of the last transaction is 172792 seconds. [172792 seconds/(60 * 60) seconds =  48 hours]. So it basically means we have a full 2 days data from the 1st point of observation. 

Since the Fraudulent transactions in general are less if we compare them to legit transactions happening at any point of time. It must be a highly unbalanced dataset.

In [6]:
credit_card_data.info()   ## Extractting data info

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 284807 entries, 0 to 284806
Data columns (total 31 columns):
 #   Column  Non-Null Count   Dtype  
---  ------  --------------   -----  
 0   Time    284807 non-null  float64
 1   V1      284807 non-null  float64
 2   V2      284807 non-null  float64
 3   V3      284807 non-null  float64
 4   V4      284807 non-null  float64
 5   V5      284807 non-null  float64
 6   V6      284807 non-null  float64
 7   V7      284807 non-null  float64
 8   V8      284807 non-null  float64
 9   V9      284807 non-null  float64
 10  V10     284807 non-null  float64
 11  V11     284807 non-null  float64
 12  V12     284807 non-null  float64
 13  V13     284807 non-null  float64
 14  V14     284807 non-null  float64
 15  V15     284807 non-null  float64
 16  V16     284807 non-null  float64
 17  V17     284807 non-null  float64
 18  V18     284807 non-null  float64
 19  V19     284807 non-null  float64
 20  V20     284807 non-null  float64
 21  V21     28

In [7]:
## Checking number of missing values in each column.

credit_card_data.isnull().sum()

Time      0
V1        0
V2        0
V3        0
V4        0
V5        0
V6        0
V7        0
V8        0
V9        0
V10       0
V11       0
V12       0
V13       0
V14       0
V15       0
V16       0
V17       0
V18       0
V19       0
V20       0
V21       0
V22       0
V23       0
V24       0
V25       0
V26       0
V27       0
V28       0
Amount    0
Class     0
dtype: int64

In [ ]:
## We see that we dont have any missing value.

In [8]:
## Checking the distribution of legit and fraudulent transaction

credit_card_data["Class"].value_counts()

0    284315
1       492
Name: Class, dtype: int64

In [ ]:
""" The above info means that in the dataset of 284806 entries, only 492 were fraudulent transaction.
    This is a highly unbalanced dataset as more than 99% of data belongs to a particular class(which in this case is 0)
    (284315/284806)*100 = 99.82% """

""" We will not feed this data to our machine learning model because it will not be able to recognise 
 any fraudulent data considering the high magnitude of legit data. """

In [9]:
## Separating the data for analysis.

legit = credit_card_data[credit_card_data.Class == 0]

fraud = credit_card_data[credit_card_data.Class == 1]

In [10]:
print(legit.shape)
print(fraud.shape)

(284315, 31)
(492, 31)


In [11]:
## statistical measures of the data

legit.Amount.describe()

count    284315.000000
mean         88.291022
std         250.105092
min           0.000000
25%           5.650000
50%          22.000000
75%          77.050000
max       25691.160000
Name: Amount, dtype: float64

In [ ]:
""" The above info gives some insights about the Amount spent in the legit transaction """

In [12]:
fraud.Amount.describe()

count     492.000000
mean      122.211321
std       256.683288
min         0.000000
25%         1.000000
50%         9.250000
75%       105.890000
max      2125.870000
Name: Amount, dtype: float64

In [ ]:
""" Of all the fraudulent activities the mean transaction if very high compared to legit transaction mean """

In [13]:
## comparing the values for both transactions

credit_card_data.groupby("Class").mean()

Time        V1        V2        V3        V4        V5  \
Class                                                                   
0      94838.202258  0.008258 -0.006271  0.012171 -0.007860  0.005453   
1      80746.806911 -4.771948  3.623778 -7.033281  4.542029 -3.151225   

             V6        V7        V8        V9  ...       V20       V21  \
Class                                          ...                       
0      0.002419  0.009637 -0.000987  0.004467  ... -0.000644 -0.001235   
1     -1.397737 -5.568731  0.570636 -2.581123  ...  0.372319  0.713588   

            V22       V23       V24       V25       V26       V27       V28  \
Class                                                                         
0     -0.000024  0.000070  0.000182 -0.000072 -0.000089 -0.000295 -0.000131   
1      0.014049 -0.040308 -0.105130  0.041449  0.051648  0.170575  0.075667   

           Amount  
Class              
0       88.291022  
1      122.211321  

[2 rows x 30 columns]

In [ ]:
""" we see a noticeable between the means of the values. This is an important clue
    This is how our Machine Learning algorithm can find difference between two types of transactions """

## Under-Sampling

Build a sample dataset containing similar distribution of normal transaction and Fraudulent transactions.

Number of Fraudulent transaction --> 492

In [ ]:
""" We will be randomly picking 492 transactions from the legit transactions
    and we will join it with the fraudulent transactions.
    This will make the new dataset a uniform and balanced dataset for analysis """

In [14]:
legit_sample = legit.sample(n = 492)   ## function sample will pick up 492 random transactions

Concatenating two DataFrames

In [17]:
new_dataset = pd.concat([legit_sample, fraud], axis = 0)    ## axis = 0 makes sure data is added in a row.

In [18]:
new_dataset.head()

Time        V1        V2        V3        V4        V5        V6  \
283017  171313.0 -0.212166  1.004959 -0.053371 -0.582689  1.326414 -0.033985   
70991    54107.0  1.277467 -0.425610  0.226724 -0.546969 -0.805828 -0.622710   
98058    66500.0  1.175519 -0.533975 -0.159998 -1.013271 -0.285744 -0.049487   
172165  120986.0 -0.045896  0.763513 -0.532738 -0.102822  0.915536 -1.414414   
47209    43133.0 -0.298882  0.795366  1.980577  1.461516 -1.114323  0.401844   

              V7        V8        V9  ...       V21       V22       V23  \
283017  0.954049  0.037412 -0.691313  ... -0.251565 -0.743009 -0.107503   
70991  -0.594040 -0.084721 -0.822915  ...  0.112326  0.227776 -0.008696   
98058  -0.164453  0.055274  1.458662  ... -0.243969 -0.632171 -0.213601   
172165  0.888381 -0.230913  0.024545  ...  0.296067  1.005467 -0.102313   
47209  -1.456283 -2.501026 -0.347608  ... -1.173221  0.539106 -0.203730   

             V24       V25       V26       V27       V28  Amount  Class  
283017  0.002023 -0.402188  0.225481  0.090355  0.166104    8.99      0  
70991   0.031325  0.298576 -0.259221  0.040824  0.046941   47.85      0  
98058  -0.836026  0.701308 -0.793005  0.046449  0.014703   72.22      0  
172165 -0.152773 -0.310968 -0.132215  0.375080  0.196584   25.00      0  
47209   0.378910  1.062235 -0.087742  0.074275  0.207766   45.00      0  

[5 rows x 31 columns]

In [19]:
new_dataset.tail()

Time        V1        V2        V3        V4        V5        V6  \
279863  169142.0 -1.927883  1.125653 -4.518331  1.749293 -1.566487 -2.010494   
280143  169347.0  1.378559  1.289381 -5.004247  1.411850  0.442581 -1.326536   
280149  169351.0 -0.676143  1.126366 -2.213700  0.468308 -1.120541 -0.003346   
281144  169966.0 -3.113832  0.585864 -5.399730  1.817092 -0.840618 -2.943548   
281674  170348.0  1.991976  0.158476 -2.583441  0.408670  1.151147 -0.096695   

              V7        V8        V9  ...       V21       V22       V23  \
279863 -0.882850  0.697211 -2.064945  ...  0.778584 -0.319189  0.639419   
280143 -1.413170  0.248525 -1.127396  ...  0.370612  0.028234 -0.145640   
280149 -2.234739  1.210158 -0.652250  ...  0.751826  0.834108  0.190944   
281144 -2.208002  1.058733 -1.632333  ...  0.583276 -0.269209 -0.456108   
281674  0.223050 -0.068384  0.577829  ... -0.164350 -0.295135 -0.072173   

             V24       V25       V26       V27       V28  Amount  Class  
279863 -0.294885  0.537503  0.788395  0.292680  0.147968  390.00      1  
280143 -0.081049  0.521875  0.739467  0.389152  0.186637    0.76      1  
280149  0.032070 -0.739695  0.471111  0.385107  0.194361   77.89      1  
281144 -0.183659 -0.328168  0.606116  0.884876 -0.253700  245.00      1  
281674 -0.450261  0.313267 -0.289617  0.002988 -0.015309   42.53      1  

[5 rows x 31 columns]

In [20]:
new_dataset["Class"].value_counts()

0    492
1    492
Name: Class, dtype: int64

In [21]:
new_dataset.groupby("Class").mean()

Time        V1        V2        V3        V4        V5  \
Class                                                                   
0      93894.829268  0.077088  0.016054  0.061220 -0.080923  0.041400   
1      80746.806911 -4.771948  3.623778 -7.033281  4.542029 -3.151225   

             V6        V7        V8        V9  ...       V20       V21  \
Class                                          ...                       
0      0.075062 -0.034597 -0.009248  0.006563  ... -0.006100 -0.027704   
1     -1.397737 -5.568731  0.570636 -2.581123  ...  0.372319  0.713588   

            V22       V23      V24       V25       V26       V27       V28  \
Class                                                                        
0      0.005037  0.011116  0.00204  0.021888  0.029025  0.005748 -0.004551   
1      0.014049 -0.040308 -0.10513  0.041449  0.051648  0.170575  0.075667   

           Amount  
Class              
0       72.026626  
1      122.211321  

[2 rows x 30 columns]

In [ ]:
""" The above step is important to check whether we got good sample or a bad sample.
    In case we got a bad sample then the mean value will be very different """

Splitting the data into Features and Targets

In [ ]:
""" Target is either 0 or 1 
    Features are V1, V2.. """

In [22]:
X = new_dataset.drop(columns = "Class", axis = 1)   ## axis = 1 would mean the column would be dropped.
y = new_dataset["Class"]

In [23]:
print(X)

            Time        V1        V2        V3        V4        V5        V6  \
283017  171313.0 -0.212166  1.004959 -0.053371 -0.582689  1.326414 -0.033985   
70991    54107.0  1.277467 -0.425610  0.226724 -0.546969 -0.805828 -0.622710   
98058    66500.0  1.175519 -0.533975 -0.159998 -1.013271 -0.285744 -0.049487   
172165  120986.0 -0.045896  0.763513 -0.532738 -0.102822  0.915536 -1.414414   
47209    43133.0 -0.298882  0.795366  1.980577  1.461516 -1.114323  0.401844   
...          ...       ...       ...       ...       ...       ...       ...   
279863  169142.0 -1.927883  1.125653 -4.518331  1.749293 -1.566487 -2.010494   
280143  169347.0  1.378559  1.289381 -5.004247  1.411850  0.442581 -1.326536   
280149  169351.0 -0.676143  1.126366 -2.213700  0.468308 -1.120541 -0.003346   
281144  169966.0 -3.113832  0.585864 -5.399730  1.817092 -0.840618 -2.943548   
281674  170348.0  1.991976  0.158476 -2.583441  0.408670  1.151147 -0.096695   

              V7        V8        V9  .

In [ ]:
""" We can see that new dataset has 30 columns instead of 31 columns that were in previous column """

In [24]:
print(y)

283017    0
70991     0
98058     0
172165    0
47209     0
         ..
279863    1
280143    1
280149    1
281144    1
281674    1
Name: Class, Length: 984, dtype: int64


Split the data into Training and Testing data

In [25]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.2, stratify = y, random_state = 2) 

## test_size = 0.2 means 80% of the data is going to Training data (X_train) and 20% of the data goes to Testing data (X_test).
## same with test data

In [26]:
print(X.shape, X_train.shape, X_test.shape)

(984, 30) (787, 30) (197, 30)


## Model Training

LogisticRegression Model is generally used for binary classification problems.

In [27]:
model = LogisticRegression()

In [28]:
## Training the LogisticRegression model with Training Data.

model.fit(X_train, y_train)

LogisticRegression()

## Model Evaluation

### Accuracy Score

In [29]:
## Checking accuracy wrt Training data

X_train_prediction = model.predict(X_train)
training_data_accuracy = accuracy_score(X_train_prediction, y_train)

In [30]:
print("Accuracy with respect to Training Data : ", training_data_accuracy)

Accuracy with respect to Training Data :  0.9466327827191868


In [ ]:
""" The above result means that out of 100 entries our Model predict results with 94.66% of accuracy """ 

In [31]:
## Checking accuracy wrt Testing data

X_test_prediction = model.predict(X_test)
test_data_accuracy = accuracy_score(X_test_prediction, y_test)

In [33]:
print("Accuracy with respect to Test Data : ", test_data_accuracy)

Accuracy with respect to Test Data :  0.9390862944162437


In [ ]:
""" The accuracy on Training Data and Test Data should not be very different from each other.
    If thats the case it would mean the model is either over-fitting or under-fitting """